### PR TITLE
Problems in slip39 find()

### PR DESCRIPTION
**Pull request by andrewkozlik**
Resolves https://github.com/satoshilabs/trezor-firmware/issues/99.

I made the following changes:
- Made the crypto/slip39.c functions more robust and moved any input validation from core to slip39.c.
- Added tests for the edge cases that @invd encountered.
- Renamed `compute_mask()` in slip39.c to `slip39_word_completion_mask()` to better correspond with the naming from bip39.c. There is an unrelated Python function `compute_mask()` in keyboard_bip39.py which makes things confusing because it does something different (computes the mask for a button).
- Added a proper SLIP39 word completion test, which goes through all the words in the wordlist and checks that they can be written with the masks that are returned by `slip39_word_completion_mask()`.
- Made the `find()` function static, since it's an internal function and cleaned it up to make it more legible. 
- I still wasn't satisfied with the legibility of `find()`, so in the last commit 7ce5a2b I reworked it in a way that I find more comprehensible. It is also more efficient since it locates the relevant prefixes using binary searches instead of a bounded linear search. The only downside I can see is that it might bloat the firmware by 2kB because it replaces the list of prefixes by a map. However, when I compared the firmware size there was no difference. I didn't want to squash the changes, because if this problem pops-up later then we can rever this commit.

### CODE REVIEW

**Review by tsusanka — APPROVED** [review]
Untested ACK.

I haven't digged deep into `find_sequence` and `button_sequence_to_word` but they LGTM.

File: crypto/slip39.c [hunk 1]
```diff
@@ -65,34 +63,64 @@ bool word_index(uint16_t* index, const char* word, uint8_t word_length) {
...
+  return hi;
 }
 
 /**
  * Returns the first word matching the button sequence prefix or NULL if no
  * match is found.
  */
-const char* button_sequence_to_word(uint16_t prefix) {
-  return get_word(find(prefix, true));
+const char* button_sequence_to_word(uint16_t sequence) {
```
Comment (onvej-sl) [inline crypto/slip39.c:94]: First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

Second, the function returns a nonsense when the prefix contains the digit zero (such as `1011`). It should be ok, since the function is never called with such a prefix. Maybe it should be at least in comments.

**Review by onvej-sl — CHANGES_REQUESTED** [review]

File: crypto/slip39.c (hunk 1 above)
Comment (prusnak) [reply crypto/slip39.c:94]: > First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

That's a good catch!

Comment (andrewkozlik) [reply crypto/slip39.c:94]: > First, the definition of the function changed. The function used to return the first word in alphabetical order matching `sequence`, now it returns the word with the lowest prefix that matches the `sequence`. I guess it will change the word that is suggested to the user.

Trezor only shows the word once there is a unique match. So this makes no difference as far as I could tell.

> Ideally we could create a unit-test for this function.

There is a unit test `test_slip39_sequence_to_word`, which I originally updated in 7ce5a2b precisely because of this.

> Second, the function returns a nonsense when the prefix contains the digit zero (such as `1011`). It should be ok, since the function is never called with such a prefix. Maybe it should be at least in comments.

Good point, but this is only a symptom of a more general problem, e.g. sequence = 255, although that is also unreachable at the moment. Nevertheless fixed in 71975ea and added more unit tests in 2c65bc4.

Comment (andrewkozlik) [reply crypto/slip39.c:94]: f69800e: Also added the same unit tests to `test_slip39_word_completion_mask`.

Comment (onvej-sl) [reply crypto/slip39.c:94]: > Trezor only shows the word once there is a unique match. So this makes no difference as far as I could tell.

You are right, I was confused, because it works differently for slip39 backup and bip39 backup.

**Review by onvej-sl — APPROVED** [review]
